### DISCUSSION TITLE
gym version will affect the usage of ray[rllib]

### DISCUSSION THREAD

**Discussion by Morphlng** [praveen-palanisamy/macad-gym#77] [General]
To enable "Multi-Agent" environment training in `rllib`, you have to inherit the base class `MultiAgentEnv` from `ray.rllib.env`. Macad-gym have already done this in `multi_env.py`.

https://github.com/praveen-palanisamy/macad-gym/blob/912944cd93c4db2f704f2818d469425fe191f976/src/macad_gym/carla/multi_env.py#L237-L243

However, since `gym==0.21.0`, the environment created with `gym.make` will automatically wrapped in a Class called `OrderEnforcing`. This will break the inheritance check in rllib, causing a training session failure.

![gym_version](https://user-images.githubusercontent.com/50166287/204316168-67120425-963c-44f4-80b7-7478b748081d.png)

This is probably something `ray[rllib]` should take care of, I'm just reporting this to help if anybody have met this problem.

**Comment by Morphlng** [comment]
There are also a bunch of incompatible api usage in the code should be noted. For example, some `_private_attribute` (e.g. Box._shape) is no longer accessible. User should whether stick with `gym==0.12.6` or we would have to do a lot of job to handle these problems.

**Comment by praveen-palanisamy** [comment]
Thanks for identifying this and posting with details! Like you pointed out, the OpenAI Gym library started wrapping the env with the `OrderEnforcing` wrapper to ensure the env throws an error if `env.step(...)` is called before an initial `env.reset()`. 
For users using latest version of `gym` (`gym>=0.21.0`) with `ray[rllib]`, current workaround is to replace `env` with `env.wrapped` in the user's code. This workaround is needed not just with [MACAD-Gym](https://github.com/praveen-palanisamy/macad-gym) environments but likely with any MultiAgent Gym environment with a wrapper class. The [`ray[rllib]` team has decided not to support Gym wrappers in their MultiAgent Env implementation](https://github.com/ray-project/ray/pull/8314).
